pkg/utils/transform/transform_test.go

```diff
@@ -171,3 +168,172 @@ func assertOnResourceList(t *testing.T, items []unstructured.Unstructured, key,
+		os.Setenv("IMAGE_PIPELINE_CONTROLLER", image)
+		testData := path.Join("testdata", "test-replace-image.yaml")
+
+		manifest, err := mf.NewManifest(testData, true, nil)
+		assetNoEror(t, err)
+		err = manifest.Transform(transform.ReplaceDeploymentImages("IMAGE_PIPELINE_"))
+
+		assetNoEror(t, err)
+		assertDeployContainersHasImage(t, manifest.Resources, "controller", image)
+		assertDeployContainersHasImage(t, manifest.Resources, "sidecar", "busybox")
+	})
+
+	t.Run("of_container_args", func(t *testing.T) {
+		image := "foo.bar/image/bash"
+		os.Setenv("IMAGE_PIPELINE_-BASH", image)
+		testData := path.Join("testdata", "test-replace-image.yaml")
+
+		manifest, err := mf.NewManifest(testData, true, nil)
+		assetNoEror(t, err)
+		err = manifest.Transform(transform.ReplaceDeploymentImages("IMAGE_PIPELINE_"))
+
+		assetNoEror(t, err)
+		assertDeployContainerArgsHasImage(t, manifest.Resources, "-bash", image)
+		assertDeployContainerArgsHasImage(t, manifest.Resources, "-git", "git")
+	})
+
+	t.Run("of_task_addons", func(t *testing.T) {
+		image := "foo.bar/image/buildah"
+		os.Setenv("IMAGE_ADDONS_PARAM_BUILDER_IMAGE", image)
+		os.Setenv("IMAGE_ADDONS_push", image)
+		testData := path.Join("testdata", "test-replace-addon-image.yaml")
+
+		manifest, err := mf.NewManifest(testData, true, nil)
+		assetNoEror(t, err)
+		err = manifest.Transform(transform.ReplaceTaskImage("IMAGE_ADDONS_", "PARAM_"))
+
+		assetNoEror(t, err)
+		assertTaskImage(t, manifest.Resources, "push", image)
+		assertParamHasImage(t, manifest.Resources, "BUILDER_IMAGE", image)
+	})
+}
+
+func assetNoEror(t *testing.T, err error) {
+	t.Helper()
+
+	if err != nil {
+		t.Errorf("assertion failed; expected no error %v", err)
+	}
+}
+
+func assertEqual(t *testing.T, result []unstructured.Unstructured, expected []unstructured.Unstructured) {
+	t.Helper()
+
+	if !reflect.DeepEqual(result, expected) {
+		t.Errorf("assertion failed; not equal: expected %v, got %v", expected, result)
+	}
+}
+
+func assertDeployContainersHasImage(t *testing.T, resources []unstructured.Unstructured, name string, image string) {
+	t.Helper()
+
+	for _, resource := range resources {
+		deployment := deploymentFor(t, resource)
+		containers := deployment.Spec.Template.Spec.Containers
+
+		for _, container := range containers {
+			if container.Name != name {
+				continue
+			}
+
+			if container.Image != image {
+				t.Errorf("assertion failed; unexpected image: expected %s and got %s", image, container.Image)
+			}
+		}
+	}
+}
+
+func assertDeployContainerArgsHasImage(t *testing.T, resources []unstructured.Unstructured, arg string, image string) {
+	t.Helper()
+
+	for _, resource := range resources {
+		deployment := deploymentFor(t, resource)
+		containers := deployment.Spec.Template.Spec.Containers
+
+		for _, container := range containers {
+			if len(container.Args) == 0 {
+				continue
+			}
+
+			for a, argument := range container.Args {
+				if argument == arg && container.Args[a+1] != image {
```

Comment:
why is it `[a+1]` instead of `[a]`

Reply:
ahh, its testing the URL of the image.

e.g. args:["-bash", "quay.io/pipeline/bash"], so when it assert for for bash -> quay.io/pipeline/bash, so ~ args[a] -> args[a+1].

right?

Reply:
i see.